File: SPCA_eps_net.py
import numpy as np

p = 1000
n = 100
X = np.random.randn(n,p)

A = (1./n)*X.T.dot(X)
U,S,tmp = np.linalg.svd(A)
A = U.dot(np.diag(S).dot(np.diag(1./np.arange(1,p+1)))).dot(np.conjugate(U.T))

del p,n,X,U,S
#---------------------------------------------

epsilon = 0.1
d = 3
k = 10
p = A.shape[0]

U,S,tmp = np.linalg.svd(A)

#Vd = U[:,1:d+1].dot(np.diag(np.sqrt(S[1:d+1])))
Vd = U[:,0:d].dot(np.diag(np.sqrt(S[0:d])))
numSamples = (4./epsilon)**d

##actual algorithm
opt_x = np.zeros((p,1))
opt_v = -np.inf

#GENERATE ALL RANDOM SAMPLES BEFORE
C = np.random.randn(d,numSamples)

for i in np.arange(1,numSamples+1):
	
	#c = np.random.randn(d,1)
	#c = C[:,i-1]
	c = C[:,i-1:i]
	c = c/np.linalg.norm(c)
	a = Vd.dot(c)

	#partial argsort in numpy?
	#if partial, kth largest is p-k th smallest
	#but need indices more than partial
	I = np.argsort(a, axis=0)
	val = np.linalg.norm(a[I[-k:]]) #index backwards to get k largest

	if val > opt_v:
		opt_v = val
		opt_x = np.zeros((p,1))
		#print((opt_x[I[0:k]]).shape)
		#print((a[I[0:k]]/val).shape)
		opt_x[I[-k:]] = a[I[-k:],:]/val


print(np.linalg.norm(np.conjugate(Vd.T).dot(opt_x)))
print(np.conjugate(opt_x.T).dot(A.dot(opt_x)))

File: SPCA_eps_net.m
%-------------------------------------------------------------------------------
%% Generate an example input for sparse PCA
%% Assume that A is given to you; This is only for my demo
p = 1000; 		% number of features per sample
n = 100;			% number of samples

X = randn(n, p);	% Create random data

A = (1/n)*X'*X;	
[U, S, ~] = svd(A); 
A = U * ( S*diag(1./[1:p]) ) * U'; % this is because I want A to behave like
  								  % real input data do. Don't worry about this part
clear p n X U S
%-------------------------------------------------------------------------------

%% Set up
% Input A, epsilon, d, k

epsilon = 0.1;
d = 3;
k = 10;		% must be between 1 and p, (p is the number of features in the data)
[p, ~] = size(A);

%% Prepare input
[U, S, ~] = svd(A); 
					% This is like [U, S] = eig(A), (A is a PSD matrix).
					% Here I use SVD because this is way more stable 
					% numerically.
					% It produces all n eigenvectors of A.
					% But we do not need them all. We only need the top d.
					% By the way, the eig function is the part that Alex was 
					% saying can be done with the so called power-iteration 
					% method, which involves multiplying a matrix with a vector
					% several times. We can talk about it...

% Sort the eigenvalues to determine which are the d largest ones

% Keep the d columns of U that correspond to the d leading eigenvectors,
% scaled with the squared root of the corresponding eigenvalues.

Vd = U(:, 1:d) * S(1:d, 1:d).^0.5; % p x d matrix

numSamples = (4/epsilon)^d; % Determine number of samples

%-------------------------------------------------------------------------------
%% The actual alogorithm

opt_x = zeros(p, 1);
opt_v = -Inf;

for i = 1:numSamples  % Each iteration is independent from the previous

	c = randn(d, 1); % Generate a random d-dimensional vector
	c = c / norm(c); % normalize its length: divide each entry by Sum(c_i^2);	
	a = Vd*c; 		 % This is a long vector with dimension p
					 % "a" is essentically a random vector in the range of Vd.
				
	[~, I] = sort(abs(a), 'descend');	% I(1:k) contains the indices of the
										% k largest entries of a (by abs value)
	
	val = norm( a(I(1:k)) );

	if val > opt_v

		opt_v = val;

		% create opt_x using the location and values of the k largest (by
		% absolute value) entries of a.
		opt_x(:)= 0;
		opt_x(I(1:k)) = a(I(1:k)) / val;
	end

end

% Although in a serial implementation this would pointless, I could essentially
% create a vector x in each iteration, store all x's, then compute 
% norm(Vd'*x) for all of them, and keep the best x (the one achieving
% the highest norm);

% opt_x is what we are interested in
% We are interested in maximizing x'*A*x. 
% opt_x should be approximately the best x
% Also, norm(Vd'*opt_x) should be less than and hopefully close to opt_x'*A*opt_x
norm(Vd'*opt_x)
opt_x'*A*opt_x
